### ISSUE TITLE
Bug: [no-redundant-type-constituents] False-positive with Record intersections

### ISSUE THREAD

**Issue by silverwind** [typescript-eslint/typescript-eslint#11811]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=C4TwDgpgBAYg9nKBeKAlCBjOAnAJgHgGdhsBLAOwHMAaKAQ3JAD4BuAKFEigCE7tlYCKADIoAbwBGfAPwAuKMTJUAvu07ReALwG9%2BoyXU1yFJCpVVA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Foo = Record<string, any>;
type Bar = Foo & {bar?: string};
type Baz = Bar & {baz?: string};
//   ^ Bar is overridden by { baz?: string | undefined; } in this intersection type.
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-redundant-type-constituents": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error, expected resulting type is `{foo?: string, bar?: string, [other: string]: any}`.

### Actual Result

Error "Bar is overridden by { baz?: string | undefined; } in this intersection type.".

**Comment by neilenns** [comment]
I also started getting this error when my dependencies auto-updated overnight. It's happening on a basic type union:

```typescript
number | undefined
```

Error is:

> error  undefined is overridden by number in this union type  @typescript-eslint/no-redundant-type-constituents

**Comment by silverwind** [comment]
Note that I initially thought this to be a regression from https://github.com/typescript-eslint/typescript-eslint/pull/10744, but that appears to not be the case, as older versions also show the issue.

At least I don't see anything wrong with this code. Typescript correctly creates a intersection type from the three constituent types, so I'm not sure what the rule wants to tell me.

**Comment by neilenns** [comment]
> Note that I initially thought this to be a regression from https://github.com/typescript-eslint/typescript-eslint/pull/10744, but that appears to not be the case, as older versions also show the issue.

My dependency update is from 8.47.0 to 8.48.0 and I did not get this error until the update to 8.48.0.

**Comment by JoshuaKGoldberg** [comment]
Fixed by reverting #10744 in #11812. If another PR attempts to resolve #10744, noting that this issue's reproductions will need to be present as test cases. Thanks for trying the new plugin+ rule version and reporting so quickly!

**Comment by neilenns** [comment]
@JoshuaKGoldberg Thanks for the quick turnaround. 8.48.1 updated in my repo last night and linting passed correctly. Yay!
